File: captainhcg/go/25-reverse-nodes-in-k-group.go
func reverse(pre *ListNode, k int) *ListNode {
    checkNode := pre.Next
    for idx := 0; idx < k; idx += 1 {
        if checkNode == nil{
            return nil
        }
        checkNode=checkNode.Next
    }
    
    start, end := pre.Next, pre.Next
    for idx := 0; idx < k - 1; idx += 1{
        oldStart := start
        start, end.Next = end.Next, end.Next.Next
        pre.Next, start.Next = start, oldStart
    }
    return end
} 

func reverseKGroup(head *ListNode, k int) *ListNode {
    if k <= 1{
        return head
    }
    dummy := &ListNode{-1, head}
    for nextPre := dummy; nextPre != nil; {
        nextPre = reverse(nextPre, k)
    }
    return dummy.Next
}

File: captainhcg/py/25-reverse-nodes-in-k-group.py
class Solution(object):
    def reverseKGroup(self, head, k):
        if k <= 1:
            return head
        dummy_head = ListNode(-1)
        dummy_head.next = head
        pre_for_next = dummy_head

        def reverse_k(pre_for_next, k):
            _nd = pre_for_next.next
            for _ in xrange(k):
                if not _nd:
                    return None
                _nd = _nd.next

            tail = pre_for_next.next
            for _ in xrange(k - 1):
                to_switch = tail.next
                tail.next = to_switch.next
                to_switch.next = pre_for_next.next
                pre_for_next.next = to_switch
            return tail

        while pre_for_next:
            pre_for_next = reverse_k(pre_for_next, k)
        return dummy_head.next
